[TypeProf][InstrFDO]Implement more efficient comparison sequence for indirect-call-promotion with vtable profiles.

mingmingl-llvm opened the pull request:
Clang's `-fwhole-program-vtables` is required for this optimization to take place. If `-fwhole-program-vtables` is not enabled, this change is no-op.
    
* Function-comparison (before):

```
%vtable = load ptr, ptr %obj
%vfn = getelementptr inbounds ptr, ptr %vtable, i64 1
%func = load ptr, ptr %vfn
%cond = icmp eq ptr %func, @callee
br i1 %cond, label bb1, label bb2:

bb1:
   call @callee

bb2:
   call %func
```

* VTable-comparison (after):

```
%vtable = load ptr, ptr %obj
%cond = icmp eq ptr %vtable, @vtable-address-point
br i1 %cond, label bb1, label bb2:

bb1:
   call @callee

bb2:
  %vfn = getelementptr inbounds ptr, ptr %vtable, i64 1
  %func = load ptr, ptr %vfn
  call %func
```
    
Key changes:
1. Find out virtual calls and the vtables they come from.
    - The ICP relies on type intrinsic `llvm.type.test` to find out virtual calls and the
        compatible vtables, and relies on type metadata to find the address point for comparison.
2. ICP pass does cost-benefit analysis and compares vtable only when the number of vtables for a function candidate is within (option specified) threshold.
3. Sink the function addressing and vtable load instruction to indirect fallback.
     - The sink helper functions are simplified versions of `InstCombinerImpl::tryToSinkInstruction`. Currently debug intrinsics are not handled. Ideally `InstCombinerImpl::tryToSinkInstructionDbgValues` and `InstCombinerImpl::tryToSinkInstructionDbgVariableRecords` could be moved into Transforms/Utils/Local.cpp (or another util cpp file) to handle debug intrinsics when moving instructions across basic blocks.
4. Keep value profiles updated
     1) Update vtable value profiles after inline
     2) For either function-based comparison or vtable-based comparison,
          update both vtable and indirect call value profiles.

mingmingl-llvm committed:
[𝘀𝗽𝗿] changes to main this commit is based on

Created using spr 1.3.4

[skip ci]

mingmingl-llvm committed:
run 'git merge main' and resolve conflict

mingmingl-llvm committed:
[InstrProf] Add vtables with type metadata into symtab to look it up with GUID

mingmingl-llvm committed:
[NFC][CallPromotionUtils]Extract a helper function versionCallSiteWithCond from versionCallSite

mingmingl-llvm committed:
[CallPromotionUtils]Implement conditional indirect call promotion with vtable-based comparison

mingmingl-llvm committed:
[TypeProf][IndirectCallPromotion]Implement vtable-based transformation

mingmingl-llvm committed:
run 'git merge main'

mingmingl-llvm committed:
[TypeProf][InstrFDO]Implement more efficient comparison sequence for
indirect-call-promotion with vtable profiles.

Clang's `-fwhole-program-vtables` is required for this optimization to
take place. If `-fwhole-program-vtables` is not enabled, this change is
no-op.

Function-comparison (before):

VTable-comparison (after):

Key changes:
1. Find out virtual calls and the vtables they come from.
   - The ICP relies on type intrinsic `llvm.type.test` and
     `llvm.public.type.test` to find out virtual calls and the
     compatible vtables, and relies on type metadata to find the address
     point (offset) for comparison.
2. ICP pass does cost-benefit analysis and compares vtable only when
   both conditions are met
   1) The function addressing and vtable load can sink to indirect
      fallback, and the indirect fallback is cold block
   2) The number of vtables for a function candidate is within
      (option specified) threshold.
3. Sink the function addressing and vtable load instruction to indirect
   fallback.
   - The sink helper functions are simplified versions of
     `InstCombinerImpl::tryToSinkInstruction`.
   - The helper functions to handle debug intrinsics are copied from
     `InstCombinerImpl::tryToSinkInstructionDbgValues` and
     `InstCombinerImpl::tryToSinkInstructionDbgVariableRecords` into
     Transforms/Utils/Local.cpp. Ideally only one copy should exist
     for inst-combine, icp and other passes.
4. Keep value profiles updated
   1) Update vtable value profiles after inline
   2) For either function-based comparison or vtable-based comparison,
      update both vtable and indirect call value profiles.

david-xl commented:
Can we rely on instcombine to do the sinking?

(the patch summary has a small error -- the comparison instruction in the example should be %cond = icmp eq ptr %vtable, @vtable-address-point

minglotus-6 commented:
> Can we rely on instcombine to do the sinking?

Good question. https://gcc.godbolt.org/z/G7Paj7h37 has two examples, `instcombine` can sink the first but cannot see through the second. I filed https://github.com/llvm/llvm-project/issues/88960 to discuss this, and I think it's more reliable to interleave the instruction sink with vtable transformations. I do think it's very preferred to re-use the helper functions like `tryToSinkInstructionDbgValues` and `tryToSinkInstructionDbgVariableRecords`.

minglotus-6 commented:
> (the patch summary has a small error -- the comparison instruction in the example should be %cond = icmp eq ptr %vtable, @vtable-address-point

thanks. I corrected it.

mingmingl-llvm committed:
Changes:
1. In Transforms/Utils/Local.{h,cpp}, remove debug intrinisc helper
   functions, and make it a TODO to handle debug intrinsic when sinking
   instructions.
2. In compiler-rt test, added lines to test IR are expected.

mingmingl-llvm committed:
undo changes to clang/lib/CodeGen/CGVTbles.cpp

david-xl commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -103,30 +110,222 @@ static cl::opt<bool>
+                             cl::Hidden,
+                             cl::desc("Percentage of vtable count to compare"));
+
+static cl::opt<int> ICPNumAdditionalVTableLast(
```
Add a comment for this option.

david-xl commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -103,30 +110,222 @@ static cl::opt<bool>
+  SmallVector<MDNode *, 2> Types; // type metadata associated with a vtable.
+  VTableVar.getMetadata(LLVMContext::MD_type, Types);
+
+  for (MDNode *Type : Types)
```
why can it have more than one types?

david-xl commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -103,30 +110,222 @@ static cl::opt<bool>
+}
+
+// Returns the basic block in which `Inst` by `Use`.
+static BasicBlock *getUserBasicBlock(Instruction *Inst, unsigned int OperandNo,
```
is "Inst" used in the function?

david-xl commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -103,30 +110,222 @@ static cl::opt<bool>
+         BB->getTerminator()->getNumSuccessors() == 2 &&
+         "Caller should guarantee");
+  // Do not sink across a critical edge for simplicity.
+  if (DestBB->getUniquePredecessor() != BB)
```
this does not handle the diamond cfg case. Perhaps remove this check. Instead check if any use is a PHI. If yes, bail.

david-xl replied:
sounds good.

david-xl commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -103,30 +110,222 @@ static cl::opt<bool>
+
+  int SinkCount = 0;
+  // FIXME: Find a way to bail out of the loop.
+  for (Instruction &I :
```
does the code assume 1 instruction to be dropped?

david-xl replied:
The question is about the drop_begin(..) call, but it is correct as the last instruction is always control transfer instruction that should be skipped. Perhaps add a brief comment: try to sink all eligible instructions in OriginalBB in reverse order.

minglotus-6 commented on llvm/lib/Transforms/Utils/InlineFunction.cpp:
```diff
@@ -1967,11 +1969,23 @@ void llvm::updateProfileCallee(
     for (auto Entry : *VMap) {
       if (isa<CallInst>(Entry.first))
-        if (auto *CI = dyn_cast_or_null<CallInst>(Entry.second))
+        if (auto *CI = dyn_cast_or_null<CallInst>(Entry.second)) {
```
Makes sense to not further duplicate the code.  How about something like this?

```
if (isa<CallBase>(Entry.first)) {
  if(auto* CB = dyn_cast_or_null<CallBase>(Entry.Base)) {
    CB->updateProfWeight(...
    Instruction* VPtr = PGOIndirectCallVisitor::tryGetVTableInstruction(CB);
    if (VPtr)
      scaleProfData(VPtr...)
}
```

`CallBase` is the base class of five IR instructions [1] now.  Among them `CoroAwaitSuspendInst` [1] and `GCStatePointInst` [2] are relatively new; the rest three are well understood. Despite more derived classes of `CallBase` might be added for new cases, the code above should be future-proof since `updateProfWeight` and `scaleProfData` are programmed to `!prof` (not `CallBase`).

[1] 
https://github.com/llvm/llvm-project/blob/692ae5443b1778e138527ef55d799a4b535a36f9/llvm/lib/Transforms/Coroutines/CoroInstr.h#L85-L88
[2]
https://github.com/llvm/llvm-project/blob/692ae5443b1778e138527ef55d799a4b535a36f9/llvm/include/llvm/IR/Statepoint.h#L61-L63

mingmingl-llvm committed:
follow up on review feedback

teresajohnson commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -321,14 +746,127 @@ bool IndirectCallPromoter::processFunction(ProfileSummaryInfo *PSI) {
+  // Right now only llvm.type.test is used to find out virtual call sites.
+  // With ThinLTO and whole-program-devirtualization, llvm.type.test and
+  // llvm.public.type.test are emitted, and llvm.public.type.test is either
+  // refined to llvm.type.test or dropped before indirect-call-promotion pass.
```
Do we then need to do the below analysis on llvm.public.type.test? ICP should be after it gets converted or dropped.

david-xl replied:
ok

teresajohnson commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -240,27 +474,102 @@ IndirectCallPromoter::getPromotionCandidatesForCallSite(
+  return Iter->second;
+}
+
+Instruction *IndirectCallPromoter::computeVTableInfos(
```
There is a lot going on in this function - can you add some more description at the top describing how this works and/or some more comments before each block of code within (particularly the latter half of the function)?

teresajohnson commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -276,35 +585,151 @@ CallBase &llvm::pgo::promoteIndirectCall(CallBase &CB, Function *DirectCallee,
+    VTableGUIDCountsMap &VTableGUIDCounts) {
+  SmallVector<uint64_t, 4> PromotedFuncCount;
+  for (const auto &Candidate : Candidates) {
+    uint64_t IfCount = 0;
```
The name IfCount isn't very intuitive to me, suggest using a different name.

teresajohnson commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -276,35 +585,151 @@ CallBase &llvm::pgo::promoteIndirectCall(CallBase &CB, Function *DirectCallee,
+    for (const auto &[GUID, VTableCount] : C.VTableGUIDAndCounts)
+      SumVTableCount += VTableCount;
+
+    for (const auto &[GUID, VTableCount] : C.VTableGUIDAndCounts) {
```
Can you add a comment describing the intuition behind the way the counts are being updated here

teresajohnson reviewed (commented):
I haven't looked at the tests yet, but some comments/questions from a first pass through most of the code

david-xl commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -103,30 +110,220 @@ static cl::opt<bool>
+    return false;
+
+  // Do not sink convergent call instructions.
+  if (const auto *C = dyn_cast<CallBase>(I))
```
is there a common utility function to do the checks below or add a TODO.

minglotus-6 replied:
Added `FIXME: Move the sink eligibility check below to a utility function in Transforms/Utils/ directory.`.

The sink correctness check is a copy of the core logic in inst-combine pass (after removing [one irrelevant case](https://github.com/llvm/llvm-project/blob/1752740f4b4b752bbe2987a0de398c6f671ceb71/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp#L4696-L4698) and making the check stricter by removing [this](https://github.com/llvm/llvm-project/blob/1752740f4b4b752bbe2987a0de398c6f671ceb71/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp#L4709))

https://github.com/llvm/llvm-project/blob/1752740f4b4b752bbe2987a0de398c6f671ceb71/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp#L4684-L4726

david-xl replied:
ok

david-xl commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -103,30 +110,220 @@ static cl::opt<bool>
+    for (BasicBlock::iterator Scan = std::next(I->getIterator()),
+                              E = I->getParent()->end();
+         Scan != E; ++Scan)
+      if (Scan->mayWriteToMemory())
```
In theory, aliasing check can be done here, but probably not important in reality for the code pattern involved.

mingmingl-llvm committed:
Changes
1. Resolve review comments.
2. Handle vtable's PGO name, like what we do for indirect-call
   promotion.
   - InstrProf.h/cpp and PGOInstrumentation.cpp are modified.
3. Make use of 'MaxNumVTableAnnotations' in PGOInstrumentation.cpp

minglotus-6 commented on llvm/include/llvm/Transforms/Utils/Local.h:
```diff
@@ -316,6 +316,15 @@ void salvageDebugInfoForDbgValues(Instruction &I,
                                   ArrayRef<DbgVariableIntrinsic *> Insns,
                                   ArrayRef<DbgVariableRecord *> DPInsns);
 
+void tryToSinkInstructionDbgValues(
```
Debug intrinsics are not handled in this PR, I un-do `Local.h/cpp` changes (and overlooked this comment).

The updated compiler-rt test (instrprof-vtable-value-prof.cpp) enabled `-g` and tested IR for the first callsite.

minglotus-6 commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -2538,6 +2538,190 @@ Value *getSalvageOpsForIcmpOp(ICmpInst *Icmp, uint64_t CurrentLocOps,
   return Icmp->getOperand(0);
 }
 
+void llvm::tryToSinkInstructionDbgValues(
```
The `gep` and `load` sequence doesn't have debug values (so far) so it's not easy to test within this PR.

IMO helper functions that handles debug value along instruction sink (e.g. https://github.com/llvm/llvm-project/blob/1752740f4b4b752bbe2987a0de398c6f671ceb71/llvm/lib/Transforms/InstCombine/InstructionCombining.cpp#L4769) will help other passes (e.g., simplify-cfg, etc) that moves instruction around. And test cases can be added for these shared helper functions.

david-xl commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -276,35 +626,154 @@ CallBase &llvm::pgo::promoteIndirectCall(CallBase &CB, Function *DirectCallee,
+
+    // 'OriginalBB' is the basic block of indirect call before indirect call
+    // promotion.
+    BasicBlock *OriginalBB = CB.getParent();
```
Move this out of the loop?

david-xl replied:
ok

david-xl commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -276,35 +626,154 @@ CallBase &llvm::pgo::promoteIndirectCall(CallBase &CB, Function *DirectCallee,
+
+    PromotedFuncCount.push_back(Candidate.Count);
+
+    TotalFuncCount -= Candidate.Count;
```
Add a check to make sure TotalFuncCount does not become negative (the difference and profile precisions in icall and vtable profile may lead to it).

minglotus-6 replied:
I added `assert(TotalFuncCount >= Candidate.Count && "message")` to keep consistent with function comparison in (https://github.com/llvm/llvm-project/blob/abedb3b8356d5d56f1c575c4f7682fba2cb19787/llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp#L286-L290)

Meanwhile used `std::min` since 'TotalFuncCount' is the saturating add result per  https://github.com/llvm/llvm-project/blob/abedb3b8356d5d56f1c575c4f7682fba2cb19787/llvm/lib/ProfileData/InstrProf.cpp#L1281-L1288

mingmingl-llvm committed:
Resolve review feedback

minglotus-6 commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -321,14 +790,114 @@ bool IndirectCallPromoter::processFunction(ProfileSummaryInfo *PSI) {
+
+    RemainingVTableCount -= Candidate.Count;
+
+    int MaxNumVTable = 1;
```
I added a comment for it.

david-xl replied:
Makes sense, but the comment may need to be updated. For non-last candidates, it conservatively limits the NumVtable to be 1 while for the last candidate it is relaxed to be a value controlled by the parameter.

minglotus-6 commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -321,14 +790,114 @@ bool IndirectCallPromoter::processFunction(ProfileSummaryInfo *PSI) {
+    RemainingVTableCount -= Candidate.Count;
+
+    int MaxNumVTable = 1;
+    if (I == Candidates.size() - 1)
```
done.

mingmingl-llvm reviewed (commented):
PTAL, thanks!

david-xl commented on llvm/lib/ProfileData/InstrProf.cpp:
```diff
@@ -1393,16 +1398,27 @@ MDNode *getPGOFuncNameMetadata(const Function &F) {
-  MDNode *N = MDNode::get(C, MDString::get(C, PGOFuncName));
-  F.setMetadata(getPGOFuncNameMetadataName(), N);
+
+  // Don't created duplictaed metadata.
```
fix typo: duplicated.

david-xl commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -103,27 +114,226 @@ static cl::opt<bool>
                  cl::desc("Dump IR after transformation happens"));
 
+// Indirect call promotion pass will fall back to function-based comparison if
+// vtable-count / function-count is smaller than this threshold.
```
In what situation the ratio will be below the 0.99 threshold?

david-xl commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -276,35 +626,154 @@ CallBase &llvm::pgo::promoteIndirectCall(CallBase &CB, Function *DirectCallee,
+    for (const auto &[GUID, VTableCount] : C.VTableGUIDAndCounts) {
+      APInt APFuncCount((unsigned)128, FuncCount, false /*signed*/);
+      APFuncCount *= VTableCount;
+      VTableGUIDCounts[GUID] -= APFuncCount.udiv(SumVTableCount).getZExtValue();
```
ok

mingmingl-llvm committed:
Changes:
1. Resolve review comments
2. Use unordered_map rather than SmallDenseMap for a couple of maps.
   - unordered_map calls allocator for each element, and (Small)DenseMap
     allocate elements in batch. But DenseMap size grows aggressively
     under size 64 [1] so not memory efficient.
3. Use stable_sort when sorting <target, count> pairs by count.
4. Only update VPtr value profiles if 'EnableVTableValueProfile' is true
   and 'VPtr' has profiles.

[1] DenseMap https://github.com/llvm/llvm-project/blob/092dbfaad257885692fa64559e9eb43a5c466798/llvm/include/llvm/ADT/DenseMap.h#L849
    SmallDenseMap https://github.com/llvm/llvm-project/blob/092dbfaad257885692fa64559e9eb43a5c466798/llvm/include/llvm/ADT/DenseMap.h#L1088

mingmingl-llvm committed:
Changes
1. remove unused headers
2. use SmallDenseMap for outer map and unorderd_map for inner map (the
   latter is more memory-efficient)

mingmingl-llvm committed:
run 'git merge main'

mingmingl-llvm committed:
update after 'git merge main'

mingmingl-llvm committed:
run 'git merge main'

teresajohnson commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -103,27 +112,226 @@ static cl::opt<bool>
+         BB->getTerminator()->getNumSuccessors() == 2 &&
+         "Guaranteed by ICP transformation");
+  // Do not sink across a critical edge for simplicity.
+  if (DestBB->getUniquePredecessor() != BB)
```
It looks like tryToSinkInstruction is called repeatedly for insts all from the same BB, and for the same DestBB - consider hoisting this check out to the caller of tryToSinkInstruction (i.e. into tryToSinkInstructions).

teresajohnson commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -140,14 +348,56 @@ class IndirectCallPromoter {
+  // pointer if type profiles exist.
+  // - Populate `VTableGUIDCounts` with <vtable-guid, count> using !prof
+  // metadata attached on the vtable pointer.
+  // - For each function candidate, finds out the vtables from which it get
```
s/get/gets/

teresajohnson commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -277,35 +626,160 @@ CallBase &llvm::pgo::promoteIndirectCall(CallBase &CB, Function *DirectCallee,
+  // FIXME: When Clang `-fstrict-vtable-pointers` is enabled, a vtable might be
+  // used to load multiple virtual functions. The vtable profiles needs to be
+  // updated properly in that case (e.g, for each indirect call annotate both
+  // type profiles and function profiles in one !prof).
```
What will happen currently? I assume it means incorrect updated profile info and not a correctness issue or crash? I'm not sure what "both type profiles and function profiles in one !prof" means - aren't they carried by different `!prof`?

teresajohnson commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -322,14 +796,133 @@ bool IndirectCallPromoter::processFunction(ProfileSummaryInfo *PSI) {
   return Changed;
 }
 
+// TODO: Returns false if the function addressing and vtable load instructions
```
s/Returns/Return/

teresajohnson commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -322,14 +796,133 @@ bool IndirectCallPromoter::processFunction(ProfileSummaryInfo *PSI) {
+    // 'MaxNumVTable' limits the number of vtables to make vtable comparison
+    // profitable. Comparing multiple vtables for one function candidate will
+    // insert additional instructions on the hot path, and allowing more than
+    // one vtable for non last candidates may or may not elongates dependency
```
s/elongates/elongate/

teresajohnson replied:
"the dependency chain"

teresajohnson commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -322,14 +796,133 @@ bool IndirectCallPromoter::processFunction(ProfileSummaryInfo *PSI) {
+    if (I == CandidateSize - 1)
+      MaxNumVTable = ICPMaxNumVTableLastCandidate;
+
+    if ((int)Candidate.AddressPoints.size() > MaxNumVTable) {
```
Might be useful to have a debug or missed optimization message for this case?

teresajohnson commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -322,14 +796,133 @@ bool IndirectCallPromoter::processFunction(ProfileSummaryInfo *PSI) {
+  // If the indirect fallback is not cold, don't compare vtables.
+  if (PSI && PSI->hasProfileSummary() &&
+      !PSI->isColdCount(RemainingVTableCount))
+    return false;
```
Ditto

teresajohnson commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -277,35 +626,160 @@ CallBase &llvm::pgo::promoteIndirectCall(CallBase &CB, Function *DirectCallee,
+             << ore::NV("DirectCallee", Candidate.TargetFunction)
+             << " with count " << ore::NV("Count", Candidate.Count)
+             << " out of " << ore::NV("TotalCount", TotalFuncCount)
+             << ", compare "
```
It might be useful to list what vtable(s) were compared

teresajohnson reviewed (approved):
lgtm with some mostly comment related fixes/suggestions

mingmingl-llvm committed:
resolve review feedback

mingmingl-llvm committed:
run 'git merge main' and fix compiler-rt test

mingmingl-llvm committed:
fix unused variable

mingmingl-llvm committed:
add back a variable which is only used by assert

mingmingl-llvm committed:
Changes:
1. Make vtable order deterministic in the vtable remark message.
2. s/Returns/Return for function-level comment.
3. Minor fixes around comment and asserts.

mingmingl-llvm committed:
run 'git merge main' and resolve conflicts

minglotus-6 commented:
Going to merge as the failed tests are not related to this PR.

```
cat github-pull-requests_build_76910_linux-linux-x64.log | grep -B 2 -A 10 "Failed Tests "
--------------------------------------------------------------------------
********************
Failed Tests (1):
  BOLT :: X86/reader-stale-yaml-std.test


Testing Time: 385.93s

Total Discovered Tests: 120952
  Skipped          :     47 (0.04%)
  Unsupported      :   3335 (2.76%)
  Passed           : 117251 (96.94%)
  Expectedly Failed:    318 (0.26%)
--
--------------------------------------------------------------------------
********************
Failed Tests (1):
  BOLT :: X86/reader-stale-yaml-std.test


Testing Time: 1.98s

Total Discovered Tests: 454
  Skipped          :   7 (1.54%)
  Unsupported      :  13 (2.86%)
  Passed           : 431 (94.93%)
  Expectedly Failed:   2 (0.44%)
```

minglotus-6 commented:
Fixing https://lab.llvm.org/buildbot/#/builders/95/builds/672 in https://github.com/llvm/llvm-project/pull/97228
